Fix recording events intermittently missing

winstona opened the pull request:
I've run into several cases where recordings are missing for events failing with the common error:

```The media could not be loaded, either because the server or network failed or because the format is not supported.```

recording config is set as:
```
record:
  enabled: True
  expire_interval: 5
  retain:
    days: 7
    mode: motion
  events:
    retain:
      default: 14
      mode: active_objects
```

In most cases, these were due to the `object_recordings_info` queue backing up and not being processed before the recording segments were deleted. There appears to be a bug in the multiprocessing python library where in some cases a non-blocking `get()` can trigger a `queue.Empty` when the queue isn't actually empty yet. A similar long-standing issue appears here: https://bugs.python.org/issue43136

This change resolves the issue by:
- changing the queue `get()` to a blocking call with a short timeout - this ensures the queue is emptied fully before the loop ends
- adding extra errors/warnings when it is likely the loop is falling behind and may be prematurely deleting segments

This likely may also help resource constrained instances, as the bug appears to be triggered more frequently in those situations.

winstona committed:
fix queues not emptying fully by changing gets to a blocking call with short timeout

NickM-27 commented:
The recordings maintainer has been rewritten in 0.13 and so far I've not seen any reports of missing segments or issues maintaining recordings. I think it would be good to have more info on logging and other data to show in what circumstances this happening

winstona commented:
I have added a separate branch including more of the debug logging I used to narrow down this issue. see here: https://github.com/winstona/frigate/tree/fix-recording-events-missing-debug

For context, this setup is using 3 cameras with CPU detectors running off of the branch above (but reverted to non-blocking queue `get(False)` calls) with the detect streams configured for 5fps and 10s segment length. Logs are processed with Loki for aggregation and generating of graphs used below.

The first thing to look at would be the additional `QueueStatsProcessor` I've added: https://github.com/winstona/frigate/blob/fix-recording-events-missing-debug/frigate/queue_stats.py

This was a quick and dirty solution to output the queue length out of band of the normal processing flow at a consistent interval (~5s) - looking at this data we can see the `object_recordings_info` queue gets consistently backed up by 300-500 frames with peaks up to 1500 frames.

![Screenshot 2023-10-15 at 3 43 30 PM](https://github.com/blakeblackshear/frigate/assets/2192755/ddf275f2-de6f-4725-9f53-9d3a0edef526)
```
max(
    last_over_time(
        {app=~"frigate.*"} 
        |= `queue stats:` 
        | regexp "queue stats: (?P<queue>.*?): (?P<count>[0-9\\.]+)" 
        | json 
        | unwrap count
        [$__interval]
    )
) by (app,queue)
```

The maintainer loop runs every 5s, and with a 10s segment length, this means that in the current implementation we **cannot** fall behind on `object_recordings_info` queue processing by more than 20s without nearly guaranteeing mis-attribution of motion/object events in `segment_stats()` allowing segments to be prematurely deleted. In this case, the queue shouldn't be longer than 5fps\*3cameras\*20s = 300 frames, with an ideal case needing to be much better than that since the motion boxes and/or active objects must be in the `object_recordings_info` dict by the time the segment is being processed in `segment_stats()`.

When the proposed fix is applied, this queue length regularly hovers around max 70 frames (the queue hovering around 50 frames was when I had temporarily disconnected a camera) :
![Screenshot 2023-10-15 at 4 08 13 PM](https://github.com/blakeblackshear/frigate/assets/2192755/fd541bcd-ac89-4557-8218-06c5260db151)


Secondarily, I have added a debug statement immediately after the queue is emptied to see how many frames are still in the queue - we should expect this to always be zero, possibly 1 or 2 due to context switching, but instead I'm seeing an inconsistent scattering of 25-600 frames still in the queue:

![Screenshot 2023-10-17 at 6 45 18 PM](https://github.com/blakeblackshear/frigate/assets/2192755/1bef5e81-03d3-43a0-aba0-977521cccfa9)
```
max(
    last_over_time(
        {app=~"frigate.*"} 
        |~ `object_recording.* queue not empty` 
        | regexp "not empty .(?P<q_len>[0-9.]+)" 
        | json 
        | unwrap q_len
        [$__interval]
    )
) by (app)
```

however with the fix, that drops down to what we might expect (this was a slightly older implementation of the debug code that always logged the size of the queue and not only when it was >5):

![Screenshot 2023-10-17 at 6 57 28 PM](https://github.com/blakeblackshear/frigate/assets/2192755/77d38ea5-c0e9-43c3-a825-38d8f7407801)


Lastly, I have added a query that logs the number of events missing recording segments here: https://github.com/winstona/frigate/blob/32c1b35a85b9439a5096ca92333d363cfd1cdf53/frigate/record/cleanup.py#L178-L194

I had 1176 events with missing recordings before I reset the sqlite db and started clean a few days ago, and in that time I've already reached 40 events without video:

![Screenshot 2023-10-17 at 7 08 15 PM](https://github.com/blakeblackshear/frigate/assets/2192755/653ff710-a308-4495-ad01-9638da223d59)
```
max(
    max_over_time(
        {app=~"frigate.*"} 
        |= `Found events missing recordings` 
        | regexp "count:(?P<count>[0-9\\.]+),\\s*min:(?P<min_t>[0-9\\.]+),\\s*max:(?P<max_t>[0-9\\.]+)" 
        | json 
        | unwrap count
        [5m]
    )
) by (app)
```

![Screenshot 2023-10-17 at 7 22 21 PM](https://github.com/blakeblackshear/frigate/assets/2192755/fc97ac94-4c78-4de1-a1c9-89b2726441a3)

Let me know if there is any additional detail I can provide.

NickM-27 reviewed (approved):
Thanks for the detailed information, that is very helpful in understanding what you are seeing. 

Based on what you've shown I think this is primarily an issue for machines that are resource constrained or otherwise not able to process this info quickly. Running on cpu detectors may be contributing to that as well.

You'll need to adjust your code to get the python checks passing, then I think this looks good

winstona committed:
add extra error/warning messages when there's a possibility of missing recording segments

NickM-27 commented:
Something seems wrong here, I am now having segment retention issues on my machine which had never had problems before. 

```
2023-10-18 07:41:41.541995415  [2023-10-18 07:41:41] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for back_deck_cam. Keeping the 5 most recent segments out of 13 and discarding the rest...
2023-10-18 07:41:41.548148901  [2023-10-18 07:41:41] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for back_gate_cam. Keeping the 5 most recent segments out of 13 and discarding the rest...
2023-10-18 07:41:41.554078495  [2023-10-18 07:41:41] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for crawlspace_cam. Keeping the 5 most recent segments out of 13 and discarding the rest...
2023-10-18 07:41:41.559200125  [2023-10-18 07:41:41] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for front_cam. Keeping the 5 most recent segments out of 13 and discarding the rest...
2023-10-18 07:41:41.567792270  [2023-10-18 07:41:41] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for front_doorbell_cam. Keeping the 5 most recent segments out of 12 and discarding the rest...
2023-10-18 07:41:41.571376846  [2023-10-18 07:41:41] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for garage_cam. Keeping the 5 most recent segments out of 13 and discarding the rest...
2023-10-18 07:41:41.575533171  [2023-10-18 07:41:41] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for office_cam. Keeping the 5 most recent segments out of 13 and discarding the rest...
2023-10-18 08:03:38.656138293  [2023-10-18 08:03:38] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for back_deck_cam. Keeping the 5 most recent segments out of 8 and discarding the rest...
2023-10-18 08:03:38.658645563  [2023-10-18 08:03:38] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for back_gate_cam. Keeping the 5 most recent segments out of 8 and discarding the rest...
2023-10-18 08:03:38.660734476  [2023-10-18 08:03:38] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for crawlspace_cam. Keeping the 5 most recent segments out of 8 and discarding the rest...
2023-10-18 08:03:38.662287181  [2023-10-18 08:03:38] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for front_cam. Keeping the 5 most recent segments out of 8 and discarding the rest...
2023-10-18 08:03:38.665439240  [2023-10-18 08:03:38] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for front_doorbell_cam. Keeping the 5 most recent segments out of 8 and discarding the rest...
2023-10-18 08:03:38.667082046  [2023-10-18 08:03:38] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for garage_cam. Keeping the 5 most recent segments out of 8 and discarding the rest...
2023-10-18 08:03:38.668888105  [2023-10-18 08:03:38] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for office_cam. Keeping the 5 most recent segments out of 8 and discarding the rest...
```

deviant77 commented:
I'm seeing the same thing...
Frigate was started at 12:59:59, so it started occurring a couple of minutes after start.

```2023-10-18 13:02:21.970891243  [2023-10-18 13:02:21] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for downstairs. Keeping the 5 most recent segments out of 8 and discarding the rest...
2023-10-18 13:02:21.971111814  [2023-10-18 13:02:21] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for front_door. Keeping the 5 most recent segments out of 8 and discarding the rest...
2023-10-18 13:02:21.971392559  [2023-10-18 13:02:21] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for front_drive. Keeping the 5 most recent segments out of 8 and discarding the rest...
2023-10-18 13:02:21.978354910  [2023-10-18 13:02:21] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for front_garage. Keeping the 5 most recent segments out of 8 and discarding the rest...
2023-10-18 13:02:21.984247393  [2023-10-18 13:02:21] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for kitchen. Keeping the 5 most recent segments out of 8 and discarding the rest...
2023-10-18 13:02:21.992827710  [2023-10-18 13:02:21] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for rear_garden. Keeping the 5 most recent segments out of 8 and discarding the rest...
2023-10-18 13:02:21.992836732  [2023-10-18 13:02:21] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for rear_patio_north. Keeping the 5 most recent segments out of 8 and discarding the rest...
2023-10-18 13:02:21.992844593  [2023-10-18 13:02:21] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for rear_patio_south. Keeping the 5 most recent segments out of 8 and discarding the rest...
2023-10-18 13:02:21.992863152  [2023-10-18 13:02:21] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for upstairs. Keeping the 5 most recent segments out of 8 and discarding the rest...
2023-10-18 13:32:05.113087100  [2023-10-18 13:32:05] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for downstairs. Keeping the 5 most recent segments out of 6 and discarding the rest...
2023-10-18 13:32:05.113095615  [2023-10-18 13:32:05] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for front_door. Keeping the 5 most recent segments out of 6 and discarding the rest...
2023-10-18 13:32:05.113206688  [2023-10-18 13:32:05] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for kitchen. Keeping the 5 most recent segments out of 6 and discarding the rest...
2023-10-18 13:32:05.113747941  [2023-10-18 13:32:05] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for rear_garden. Keeping the 5 most recent segments out of 6 and discarding the rest...
2023-10-18 13:32:05.117203887  [2023-10-18 13:32:05] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for rear_patio_south. Keeping the 5 most recent segments out of 6 and discarding the rest...
2023-10-18 13:32:05.118860263  [2023-10-18 13:32:05] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for upstairs. Keeping the 5 most recent segments out of 6 and discarding the rest...
2023-10-18 13:37:30.189739620  [2023-10-18 13:37:30] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for downstairs. Keeping the 5 most recent segments out of 6 and discarding the rest...
2023-10-18 13:37:30.190664847  [2023-10-18 13:37:30] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for front_door. Keeping the 5 most recent segments out of 6 and discarding the rest...
2023-10-18 13:37:30.190774141  [2023-10-18 13:37:30] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for front_drive. Keeping the 5 most recent segments out of 7 and discarding the rest...
2023-10-18 13:37:30.194039839  [2023-10-18 13:37:30] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for front_garage. Keeping the 5 most recent segments out of 6 and discarding the rest...
2023-10-18 13:37:30.194474303  [2023-10-18 13:37:30] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for kitchen. Keeping the 5 most recent segments out of 6 and discarding the rest...
2023-10-18 13:37:30.194479301  [2023-10-18 13:37:30] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for rear_garden. Keeping the 5 most recent segments out of 6 and discarding the rest...
2023-10-18 13:37:30.194709293  [2023-10-18 13:37:30] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for rear_patio_north. Keeping the 5 most recent segments out of 6 and discarding the rest...
2023-10-18 13:37:30.195059332  [2023-10-18 13:37:30] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for rear_patio_south. Keeping the 5 most recent segments out of 6 and discarding the rest...
2023-10-18 13:37:30.195410955  [2023-10-18 13:37:30] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for upstairs. Keeping the 5 most recent segments out of 6 and discarding the rest...
2023-10-18 13:40:13.636979695  [2023-10-18 13:40:13] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for downstairs. Keeping the 5 most recent segments out of 6 and discarding the rest...
2023-10-18 13:40:13.637699169  [2023-10-18 13:40:13] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for front_door. Keeping the 5 most recent segments out of 6 and discarding the rest...
2023-10-18 13:40:13.637705666  [2023-10-18 13:40:13] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for front_garage. Keeping the 5 most recent segments out of 6 and discarding the rest...
2023-10-18 13:40:13.638148985  [2023-10-18 13:40:13] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for kitchen. Keeping the 5 most recent segments out of 6 and discarding the rest...
2023-10-18 13:40:13.639113097  [2023-10-18 13:40:13] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for rear_garden. Keeping the 5 most recent segments out of 6 and discarding the rest...
2023-10-18 13:40:13.639120552  [2023-10-18 13:40:13] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for rear_patio_north. Keeping the 5 most recent segments out of 6 and discarding the rest...
2023-10-18 13:40:13.642152401  [2023-10-18 13:40:13] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for rear_patio_south. Keeping the 5 most recent segments out of 6 and discarding the rest...
2023-10-18 13:40:13.643773768  [2023-10-18 13:40:13] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for upstairs. Keeping the 5 most recent segments out of 6 and discarding the rest...
2023-10-18 13:41:52.644448621  [2023-10-18 13:41:52] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for downstairs. Keeping the 5 most recent segments out of 7 and discarding the rest...
2023-10-18 13:41:52.645457780  [2023-10-18 13:41:52] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for front_door. Keeping the 5 most recent segments out of 7 and discarding the rest...
2023-10-18 13:41:52.645765180  [2023-10-18 13:41:52] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for front_drive. Keeping the 5 most recent segments out of 7 and discarding the rest...
2023-10-18 13:41:52.646970121  [2023-10-18 13:41:52] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for front_garage. Keeping the 5 most recent segments out of 7 and discarding the rest...
2023-10-18 13:41:52.647650755  [2023-10-18 13:41:52] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for kitchen. Keeping the 5 most recent segments out of 7 and discarding the rest...
2023-10-18 13:41:52.648744806  [2023-10-18 13:41:52] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for rear_garden. Keeping the 5 most recent segments out of 7 and discarding the rest...
2023-10-18 13:41:52.650410129  [2023-10-18 13:41:52] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for rear_patio_north. Keeping the 5 most recent segments out of 7 and discarding the rest...
2023-10-18 13:41:52.651485038  [2023-10-18 13:41:52] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for rear_patio_south. Keeping the 5 most recent segments out of 7 and discarding the rest...
2023-10-18 13:41:52.652855837  [2023-10-18 13:41:52] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for upstairs. Keeping the 5 most recent segments out of 7 and discarding the rest...
2023-10-18 13:43:06.420951651  [2023-10-18 13:43:06] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for downstairs. Keeping the 5 most recent segments out of 6 and discarding the rest...
2023-10-18 13:43:06.420960642  [2023-10-18 13:43:06] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for front_door. Keeping the 5 most recent segments out of 7 and discarding the rest...
2023-10-18 13:43:06.421970870  [2023-10-18 13:43:06] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for front_drive. Keeping the 5 most recent segments out of 6 and discarding the rest...
2023-10-18 13:43:06.423404515  [2023-10-18 13:43:06] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for front_garage. Keeping the 5 most recent segments out of 6 and discarding the rest...
2023-10-18 13:43:06.424231874  [2023-10-18 13:43:06] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for kitchen. Keeping the 5 most recent segments out of 6 and discarding the rest...
2023-10-18 13:43:06.428001540  [2023-10-18 13:43:06] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for rear_garden. Keeping the 5 most recent segments out of 6 and discarding the rest...
2023-10-18 13:43:06.428010113  [2023-10-18 13:43:06] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for rear_patio_north. Keeping the 5 most recent segments out of 6 and discarding the rest...
2023-10-18 13:43:06.428013578  [2023-10-18 13:43:06] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for rear_patio_south. Keeping the 5 most recent segments out of 6 and discarding the rest...
2023-10-18 13:43:06.429239708  [2023-10-18 13:43:06] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for upstairs. Keeping the 5 most recent segments out of 6 and discarding the rest...
2023-10-18 13:48:43.601674388  [2023-10-18 13:48:43] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for downstairs. Keeping the 5 most recent segments out of 12 and discarding the rest...
2023-10-18 13:48:43.603152874  [2023-10-18 13:48:43] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for front_door. Keeping the 5 most recent segments out of 12 and discarding the rest...
2023-10-18 13:48:43.604881339  [2023-10-18 13:48:43] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for front_drive. Keeping the 5 most recent segments out of 11 and discarding the rest...
2023-10-18 13:48:43.608151931  [2023-10-18 13:48:43] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for front_garage. Keeping the 5 most recent segments out of 12 and discarding the rest...
2023-10-18 13:48:43.609456591  [2023-10-18 13:48:43] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for kitchen. Keeping the 5 most recent segments out of 12 and discarding the rest...
2023-10-18 13:48:43.610555980  [2023-10-18 13:48:43] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for rear_garden. Keeping the 5 most recent segments out of 12 and discarding the rest...
2023-10-18 13:48:43.615644368  [2023-10-18 13:48:43] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for rear_patio_north. Keeping the 5 most recent segments out of 11 and discarding the rest...
2023-10-18 13:48:43.619981456  [2023-10-18 13:48:43] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for rear_patio_south. Keeping the 5 most recent segments out of 12 and discarding the rest...
2023-10-18 13:48:43.622901430  [2023-10-18 13:48:43] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for upstairs. Keeping the 5 most recent segments out of 11 and discarding the rest...
2023-10-18 13:50:11.392353275  [2023-10-18 13:50:11] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for downstairs. Keeping the 5 most recent segments out of 8 and discarding the rest...
2023-10-18 13:50:11.394460517  [2023-10-18 13:50:11] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for front_door. Keeping the 5 most recent segments out of 8 and discarding the rest...
2023-10-18 13:50:11.395096406  [2023-10-18 13:50:11] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for front_drive. Keeping the 5 most recent segments out of 9 and discarding the rest...
2023-10-18 13:50:11.398028454  [2023-10-18 13:50:11] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for front_garage. Keeping the 5 most recent segments out of 8 and discarding the rest...
2023-10-18 13:50:11.398549801  [2023-10-18 13:50:11] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for kitchen. Keeping the 5 most recent segments out of 8 and discarding the rest...
2023-10-18 13:50:11.398807486  [2023-10-18 13:50:11] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for rear_garden. Keeping the 5 most recent segments out of 8 and discarding the rest...
2023-10-18 13:50:11.400785610  [2023-10-18 13:50:11] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for rear_patio_north. Keeping the 5 most recent segments out of 8 and discarding the rest...
2023-10-18 13:50:11.401528297  [2023-10-18 13:50:11] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for rear_patio_south. Keeping the 5 most recent segments out of 8 and discarding the rest...
2023-10-18 13:50:11.407087811  [2023-10-18 13:50:11] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for upstairs. Keeping the 5 most recent segments out of 9 and discarding the rest...
2023-10-18 13:52:29.029461535  [2023-10-18 13:52:29] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for downstairs. Keeping the 5 most recent segments out of 9 and discarding the rest...
2023-10-18 13:52:29.030524292  [2023-10-18 13:52:29] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for front_door. Keeping the 5 most recent segments out of 9 and discarding the rest...
2023-10-18 13:52:29.031203751  [2023-10-18 13:52:29] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for front_drive. Keeping the 5 most recent segments out of 9 and discarding the rest...
2023-10-18 13:52:29.033764698  [2023-10-18 13:52:29] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for front_garage. Keeping the 5 most recent segments out of 9 and discarding the rest...
2023-10-18 13:52:29.034947733  [2023-10-18 13:52:29] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for kitchen. Keeping the 5 most recent segments out of 9 and discarding the rest...
2023-10-18 13:52:29.035591765  [2023-10-18 13:52:29] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for rear_garden. Keeping the 5 most recent segments out of 9 and discarding the rest...
2023-10-18 13:52:29.042487641  [2023-10-18 13:52:29] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for rear_patio_north. Keeping the 5 most recent segments out of 9 and discarding the rest...
2023-10-18 13:52:29.045519487  [2023-10-18 13:52:29] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for rear_patio_south. Keeping the 5 most recent segments out of 9 and discarding the rest...
2023-10-18 13:52:29.047480507  [2023-10-18 13:52:29] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for upstairs. Keeping the 5 most recent segments out of 9 and discarding the rest...
2023-10-18 13:53:37.554386400  [2023-10-18 13:53:37] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for downstairs. Keeping the 5 most recent segments out of 7 and discarding the rest...
2023-10-18 13:53:37.557213832  [2023-10-18 13:53:37] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for front_door. Keeping the 5 most recent segments out of 7 and discarding the rest...
2023-10-18 13:53:37.558423660  [2023-10-18 13:53:37] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for front_drive. Keeping the 5 most recent segments out of 7 and discarding the rest...
2023-10-18 13:53:37.559249943  [2023-10-18 13:53:37] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for front_garage. Keeping the 5 most recent segments out of 7 and discarding the rest...
2023-10-18 13:53:37.559623257  [2023-10-18 13:53:37] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for kitchen. Keeping the 5 most recent segments out of 7 and discarding the rest...
2023-10-18 13:53:37.559923172  [2023-10-18 13:53:37] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for rear_garden. Keeping the 5 most recent segments out of 7 and discarding the rest...
2023-10-18 13:53:37.560760653  [2023-10-18 13:53:37] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for rear_patio_north. Keeping the 5 most recent segments out of 7 and discarding the rest...
2023-10-18 13:53:37.561836587  [2023-10-18 13:53:37] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for rear_patio_south. Keeping the 5 most recent segments out of 7 and discarding the rest...
2023-10-18 13:53:37.562736239  [2023-10-18 13:53:37] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for upstairs. Keeping the 5 most recent segments out of 7 and discarding the rest...
2023-10-18 13:54:50.781567319  [2023-10-18 13:54:50] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for downstairs. Keeping the 5 most recent segments out of 7 and discarding the rest...
2023-10-18 13:54:50.782348470  [2023-10-18 13:54:50] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for front_door. Keeping the 5 most recent segments out of 7 and discarding the rest...
2023-10-18 13:54:50.782604452  [2023-10-18 13:54:50] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for front_drive. Keeping the 5 most recent segments out of 8 and discarding the rest...
2023-10-18 13:54:50.787599265  [2023-10-18 13:54:50] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for front_garage. Keeping the 5 most recent segments out of 7 and discarding the rest...
2023-10-18 13:54:50.789091865  [2023-10-18 13:54:50] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for kitchen. Keeping the 5 most recent segments out of 7 and discarding the rest...
2023-10-18 13:54:50.789843452  [2023-10-18 13:54:50] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for rear_garden. Keeping the 5 most recent segments out of 7 and discarding the rest...
2023-10-18 13:54:50.790639331  [2023-10-18 13:54:50] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for rear_patio_north. Keeping the 5 most recent segments out of 7 and discarding the rest...
2023-10-18 13:54:50.790977790  [2023-10-18 13:54:50] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for rear_patio_south. Keeping the 5 most recent segments out of 7 and discarding the rest...
2023-10-18 13:54:50.792061891  [2023-10-18 13:54:50] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for upstairs. Keeping the 5 most recent segments out of 7 and discarding the rest...
2023-10-18 13:56:17.521233182  [2023-10-18 13:56:17] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for downstairs. Keeping the 5 most recent segments out of 6 and discarding the rest...
2023-10-18 13:56:17.521509966  [2023-10-18 13:56:17] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for front_door. Keeping the 5 most recent segments out of 6 and discarding the rest...
2023-10-18 13:56:17.522200124  [2023-10-18 13:56:17] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for front_garage. Keeping the 5 most recent segments out of 6 and discarding the rest...
2023-10-18 13:56:17.522456515  [2023-10-18 13:56:17] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for kitchen. Keeping the 5 most recent segments out of 6 and discarding the rest...
2023-10-18 13:56:17.522626439  [2023-10-18 13:56:17] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for rear_garden. Keeping the 5 most recent segments out of 6 and discarding the rest...
2023-10-18 13:56:17.522735223  [2023-10-18 13:56:17] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for rear_patio_north. Keeping the 5 most recent segments out of 6 and discarding the rest...
2023-10-18 13:56:17.522978843  [2023-10-18 13:56:17] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for rear_patio_south. Keeping the 5 most recent segments out of 6 and discarding the rest...
2023-10-18 13:56:17.528910071  [2023-10-18 13:56:17] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for upstairs. Keeping the 5 most recent segments out of 6 and discarding the rest...
2023-10-18 14:33:16.009439319  [2023-10-18 14:33:16] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for downstairs. Keeping the 5 most recent segments out of 8 and discarding the rest...
2023-10-18 14:33:16.017648157  [2023-10-18 14:33:16] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for front_door. Keeping the 5 most recent segments out of 8 and discarding the rest...
2023-10-18 14:33:16.019800717  [2023-10-18 14:33:16] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for front_drive. Keeping the 5 most recent segments out of 7 and discarding the rest...
2023-10-18 14:33:16.022400622  [2023-10-18 14:33:16] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for front_garage. Keeping the 5 most recent segments out of 8 and discarding the rest...
2023-10-18 14:33:16.029209713  [2023-10-18 14:33:16] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for kitchen. Keeping the 5 most recent segments out of 8 and discarding the rest...
2023-10-18 14:33:16.032195307  [2023-10-18 14:33:16] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for rear_garden. Keeping the 5 most recent segments out of 8 and discarding the rest...
2023-10-18 14:33:16.042729538  [2023-10-18 14:33:16] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for rear_patio_north. Keeping the 5 most recent segments out of 8 and discarding the rest...
2023-10-18 14:33:16.045046346  [2023-10-18 14:33:16] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for rear_patio_south. Keeping the 5 most recent segments out of 8 and discarding the rest...
2023-10-18 14:33:16.056061783  [2023-10-18 14:33:16] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for upstairs. Keeping the 5 most recent segments out of 8 and discarding the rest...
2023-10-18 14:36:59.715966526  [2023-10-18 14:36:59] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for downstairs. Keeping the 5 most recent segments out of 8 and discarding the rest...
2023-10-18 14:36:59.716547569  [2023-10-18 14:36:59] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for front_door. Keeping the 5 most recent segments out of 8 and discarding the rest...
2023-10-18 14:36:59.717968905  [2023-10-18 14:36:59] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for front_drive. Keeping the 5 most recent segments out of 8 and discarding the rest...
2023-10-18 14:36:59.720291791  [2023-10-18 14:36:59] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for front_garage. Keeping the 5 most recent segments out of 8 and discarding the rest...
2023-10-18 14:36:59.721728823  [2023-10-18 14:36:59] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for kitchen. Keeping the 5 most recent segments out of 8 and discarding the rest...
2023-10-18 14:36:59.722211352  [2023-10-18 14:36:59] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for rear_garden. Keeping the 5 most recent segments out of 8 and discarding the rest...
2023-10-18 14:36:59.724377909  [2023-10-18 14:36:59] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for rear_patio_north. Keeping the 5 most recent segments out of 8 and discarding the rest...
2023-10-18 14:36:59.727500959  [2023-10-18 14:36:59] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for rear_patio_south. Keeping the 5 most recent segments out of 8 and discarding the rest...
2023-10-18 14:36:59.728723245  [2023-10-18 14:36:59] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for upstairs. Keeping the 5 most recent segments out of 8 and discarding the rest...
2023-10-18 14:39:03.271515540  [2023-10-18 14:39:03] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for downstairs. Keeping the 5 most recent segments out of 7 and discarding the rest...
2023-10-18 14:39:03.272530615  [2023-10-18 14:39:03] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for front_door. Keeping the 5 most recent segments out of 6 and discarding the rest...
2023-10-18 14:39:03.274342295  [2023-10-18 14:39:03] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for front_drive. Keeping the 5 most recent segments out of 6 and discarding the rest...
2023-10-18 14:39:03.282546347  [2023-10-18 14:39:03] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for front_garage. Keeping the 5 most recent segments out of 7 and discarding the rest...
2023-10-18 14:39:03.299058143  [2023-10-18 14:39:03] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for kitchen. Keeping the 5 most recent segments out of 7 and discarding the rest...
2023-10-18 14:39:03.300487253  [2023-10-18 14:39:03] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for rear_garden. Keeping the 5 most recent segments out of 7 and discarding the rest...
2023-10-18 14:39:03.308081682  [2023-10-18 14:39:03] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for rear_patio_north. Keeping the 5 most recent segments out of 6 and discarding the rest...
2023-10-18 14:39:03.312579638  [2023-10-18 14:39:03] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for rear_patio_south. Keeping the 5 most recent segments out of 7 and discarding the rest...
2023-10-18 14:39:03.325611568  [2023-10-18 14:39:03] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for upstairs. Keeping the 5 most recent segments out of 6 and discarding the rest...
2023-10-18 14:40:00.504963160  [2023-10-18 14:40:00] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for front_door. Keeping the 5 most recent segments out of 6 and discarding the rest...
2023-10-18 14:40:00.505625819  [2023-10-18 14:40:00] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for upstairs. Keeping the 5 most recent segments out of 6 and discarding the rest...
2023-10-18 14:44:49.643376800  [2023-10-18 14:44:49] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for downstairs. Keeping the 5 most recent segments out of 14 and discarding the rest...
2023-10-18 14:44:49.644796493  [2023-10-18 14:44:49] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for front_door. Keeping the 5 most recent segments out of 14 and discarding the rest...
2023-10-18 14:44:49.646820125  [2023-10-18 14:44:49] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for front_drive. Keeping the 5 most recent segments out of 13 and discarding the rest...
2023-10-18 14:44:49.652339186  [2023-10-18 14:44:49] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for front_garage. Keeping the 5 most recent segments out of 14 and discarding the rest...
2023-10-18 14:44:49.654373357  [2023-10-18 14:44:49] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for kitchen. Keeping the 5 most recent segments out of 14 and discarding the rest...
2023-10-18 14:44:49.655492483  [2023-10-18 14:44:49] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for rear_garden. Keeping the 5 most recent segments out of 14 and discarding the rest...
2023-10-18 14:44:49.659562728  [2023-10-18 14:44:49] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for rear_patio_north. Keeping the 5 most recent segments out of 14 and discarding the rest...
2023-10-18 14:44:49.667772845  [2023-10-18 14:44:49] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for rear_patio_south. Keeping the 5 most recent segments out of 14 and discarding the rest...
2023-10-18 14:44:49.671472890  [2023-10-18 14:44:49] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for upstairs. Keeping the 5 most recent segments out of 13 and discarding the rest...
2023-10-18 14:48:04.099276326  [2023-10-18 14:48:04] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for downstairs. Keeping the 5 most recent segments out of 6 and discarding the rest...
2023-10-18 14:48:04.099417893  [2023-10-18 14:48:04] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for front_door. Keeping the 5 most recent segments out of 6 and discarding the rest...
2023-10-18 14:48:04.099632005  [2023-10-18 14:48:04] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for front_drive. Keeping the 5 most recent segments out of 6 and discarding the rest...
2023-10-18 14:48:04.100652967  [2023-10-18 14:48:04] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for front_garage. Keeping the 5 most recent segments out of 6 and discarding the rest...
2023-10-18 14:48:04.100660951  [2023-10-18 14:48:04] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for kitchen. Keeping the 5 most recent segments out of 6 and discarding the rest...
2023-10-18 14:48:04.101120819  [2023-10-18 14:48:04] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for rear_garden. Keeping the 5 most recent segments out of 6 and discarding the rest...
2023-10-18 14:48:04.101413461  [2023-10-18 14:48:04] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for rear_patio_north. Keeping the 5 most recent segments out of 6 and discarding the rest...
2023-10-18 14:48:04.101810362  [2023-10-18 14:48:04] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for rear_patio_south. Keeping the 5 most recent segments out of 6 and discarding the rest...
2023-10-18 14:48:04.102011555  [2023-10-18 14:48:04] frigate.record.maintainer      WARNING : Unable to keep up with recording segments in cache for upstairs. Keeping the 5 most recent segments out of 6 and discarding the rest...
2023-10-18 15:00:10.129629407  [2023-10-18 15:00:10] frigate.record.maintainer      ERROR   : Found 1 old frames, segments from recordings may be missing```

winstona commented:
thanks for the quick patch @NickM-27 to resolve this, I see why that issue could happen now, and had overlooked those consequences. I'm just verifying now but I'm sure the `timeout=0.01` will also be sufficient enough in my case.

This may not be necessary at this point, but a couple ideas on how to increase the robustness of this loop:
1. add a deadline to the queue `get(..)` loop to break out and ensure `move_files()` gets run at least once every `x` seconds
2. delay calling `move_segment()` until `object_recordings_info[camera]` has stats at least `y` seconds after the segments `end_time` - possibly allowing `move_files()` to be run significantly more frequently without the risk of not having stats available. This does assume that the queue gets frames in order, however I believe that is not only the case, but an assumption that is already being made.

I can code something up if any of these (or any other ideas) are worth pursuing
